packages/dds/legacy-dds/src/sharedSignal.ts

```diff
@@ -0,0 +1,169 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import type {
+	IChannelAttributes,
+	IFluidDataStoreRuntime,
+	IChannelStorageService,
+	IChannelFactory,
+} from "@fluidframework/datastore-definitions/internal";
+import { FileMode, MessageType, TreeEntry } from "@fluidframework/driver-definitions/internal";
+import type {
+	ISequencedDocumentMessage,
+	ITree,
+} from "@fluidframework/driver-definitions/internal";
+import type { ISummaryTreeWithStats } from "@fluidframework/runtime-definitions/internal";
+import { convertToSummaryTreeWithStats } from "@fluidframework/runtime-utils/internal";
+import type { IFluidSerializer } from "@fluidframework/shared-object-base/internal";
+import { SharedObject } from "@fluidframework/shared-object-base/internal";
+
+import type {
+	ISharedSignal,
+	ISharedSignalEvents,
+	ISignalOperation,
+	SerializableTypeForSharedSignal,
+} from "./interfaces.js";
+import { SharedSignalFactory } from "./sharedSignalFactory.js";
+
+const snapshotFileName = "header";
+
+/**
+ * Represents a shared signal that allows communication between distributed clients.
+ *
+ * @internal
+ */
+export class SharedSignal<T extends SerializableTypeForSharedSignal = any>
+	extends SharedObject<ISharedSignalEvents<T>>
+	implements ISharedSignal<T>
+{
+	/**
+	 * Create a new shared signal
+	 *
+	 * @param runtime - data store runtime the new shared signal belongs to
+	 * @param id - optional name of the shared signal
+	 * @returns newly create shared signal (but not attached yet)
+	 */
+	public static create(runtime: IFluidDataStoreRuntime, id?: string): SharedSignal {
+		return runtime.createChannel(id, SharedSignalFactory.Type) as SharedSignal;
+	}
+
+	/**
+	 * Get a factory for SharedSignal to register with the data store.
+	 *
+	 * @returns a factory that creates and load SharedSignal
+	 */
+	public static getFactory(): IChannelFactory {
+		return new SharedSignalFactory();
+	}
+
+	/**
+	 * Constructs a new shared signal. If the object is non-local an id and service interfaces will
+	 * be provided
+	 * @param id - optional name of the shared signal
+	 * @param runtime - data store runtime the shared signal belongs to
+	 * @param attributes - represents the attributes of a channel/DDS.
+	 */
+	public constructor(
+		id: string,
+		runtime: IFluidDataStoreRuntime,
+		attributes: IChannelAttributes,
+	) {
+		super(id, runtime, attributes, "loop_sharedSignal_" /* telemetryContextPrefix */);
+	}
+
+	/**
+	 * Method used for generating a signal.
+	 */
+	public notify(metadata?: T): void {
+		// If we are not attached, don't submit the op.
+		if (!this.isAttached()) {
+			return;
+		}
+
+		const op: ISignalOperation<T> = {
+			type: "signal",
+			metadata,
+		};
+
+		this.notifyCore(op, true);
```

Comment:
Looks like this is going to emit an event twice for local ops - once when sending and once when receiving the response.  Not for this PR, but would be interesting to investigate in a followup.